ghost/email-service/lib/mailgun-email-provider.js

```diff
@@ -0,0 +1,168 @@
+ * @prop {string} email
+ * @prop {Replacement[]} replacements
+ */
+
+/**
+ * @typedef {object} Replacement
+ * @prop {string} token
+ * @prop {string} value
+ * @prop {string} id
+ */
+
+/**
+ * @typedef {object} EmailSendingOptions
+ * @prop {boolean} clickTrackingEnabled
+ * @prop {boolean} openTrackingEnabled
+ */
+
+/**
+ * @typedef {object} EmailProviderSuccessResponse
+ * @prop {string} id
+ */
+
+class MailgunEmailProvider {
+    #mailgunClient;
+    #errorHandler;
+
+    static BATCH_SIZE = 1000;
+
+    /**
+     * @param {object} dependencies
+     * @param {import('@tryghost/mailgun-client/lib/mailgun-client')} dependencies.mailgunClient - mailgun client to send emails
+     * @param {Function} [dependencies.errorHandler] - custom error handler for logging exceptions
+     */
+    constructor({
+        mailgunClient,
+        errorHandler
+    }) {
+        this.#mailgunClient = mailgunClient;
+        this.#errorHandler = errorHandler;
+    }
+
+    #createRecipientData(replacements) {
+        let recipientData = {};
+
+        recipientData = replacements.reduce((acc, replacement) => {
+            const {id, value} = replacement;
+            if (!acc[id]) {
+                acc[id] = {};
+            }
+            acc[id] = value;
+            return acc;
+        }, {});
+
+        return recipientData;
+    }
+
+    #updateRecipientVariables(data, replacementDefinitions) {
+        for (const def of replacementDefinitions) {
+            data = data.replace(
+                def.token,
+                `%recipient.${def.id}%`
+            );
+        }
+        return data;
+    }
+
+    /**
+     * Create mailgun error message for storing in the database
+     * @param {Object} error
+     * @param {string} error.message
+     * @param {string} error.details
+     * @returns {string}
+    */
+    #createMailgunErrorMessage(error) {
+        const message = (error?.message || '') + ':' + (error?.details || '');
+        return message.slice(0, 2000);
+    }
+
+    /**
+     * Send an email using the Mailgun API
+     * @param {import('./sending-service').EmailData} data
+     * @param {EmailSendingOptions} options
+     * @returns {Promise<EmailProviderSuccessResponse>}
+    */
+    async send(data, options) {
+        const {
+            subject,
+            html,
+            plaintext,
+            from,
+            replyTo,
+            id,
+            recipients,
+            replacementDefinitions
+        } = data;
+
+        logging.info(`Sending email to ${recipients.length} recipients`);
+        const startTime = Date.now();
+        debug(`sending message to ${recipients.length} recipients`);
+
+        try {
+            const messageData = {
+                subject,
+                html,
+                plaintext,
+                from,
+                replyTo,
+                id,
+                track_opens: !!options.openTrackingEnabled,
+                track_clicks: !!options.clickTrackingEnabled
```

Comment:
This should always be false because we do our own link tracking

Reply:
Maybe we can just set `clickTrackingEnabled` to false in the sending service. Because the email renderer is adding the click tracking

Reply:
Yeah i think this should come from outside of provider, because provider shouldn't have any of our app logic in it but works with the option provided. So as you said we can set `clickTrackingEnabled` as false in sending service which will automatically come down :)